### ISSUE TITLE
The module or file RescriptCore can't be found.

### ISSUE THREAD

**Issue by BIYUEHU** [rescript-lang/rescript-vscode#1056]
Report details:

```txt
The module or file RescriptCore can't be found.
- If it's a third-party dependency:
  - Did you add it to the "bs-dependencies" or "bs-dev-dependencies" in bsconfig.json?
- Did you include the file's directory to the "sources" in bsconfig.json?
```

- Extension version: `v1.58.0`
- I created a basic project by `create-rescript-app` and already installed all packages.
- There isn't non-english Characters at the project path: `E:\repos\my-res`

Part of `Package.json`:

```json
  "dependencies": {
    "@rescript/core": "^1.3.0",
    "rescript": "^11.1.0"
  }
```

`Visual Studio Code` information:

```txt
Version: 1.85.2 (user setup)
Commit: 8b3775030ed1a69b13e4f4c628c612102e30a681
Date: 2024-01-18T06:40:10.514Z(10 mos ago)
Electron: 25.9.7
ElectronBuildld:26354273
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
Os: Windows NT x64 10.0.22621
```

Screenboot:

![QQ截图20241117160728](https://github.com/user-attachments/assets/67db23af-9379-4bb2-a405-bb963bf1daed)

**Comment by nojaf** [comment]
Hello, could you share your `rescript.json`?

**Comment by BIYUEHU** [comment]
> Hello, could you share your `rescript.json`?

Sorry, I forgot. Here:

```json
{
  "name": "rescript-project-template",
  "sources": {
    "dir": "src",
    "subdirs": true
  },
  "package-specs": {
    "module": "commonjs",
    "in-source": true
  },
  "suffix": ".res.js",
  "bs-dependencies": ["@rescript/core"],
  "bsc-flags": ["-open RescriptCore"]
}
```

Everything was generated automatically through create-rescript-app, and I didn't make any changes.

**Comment by fhammerschmidt** [comment]
I just created a create-rescript-app basic project with pnpm and do not have this issue. I hope it's not windows-specific.

I guess you also tried to restart the language server from VSCode's command palette?

Other than that what does the OUTPUT tab say when you select "ReScript Language Server" from the dropdown menu?

**Comment by BIYUEHU** [comment]
> I just created a create-rescript-app basic project with pnpm and do not have this issue. I hope it's not windows-specific.
> 
> I guess you also tried to restart the language server from VSCode's command palette?
> 
> Other than that what does the OUTPUT tab say when you select "ReScript Language Server" from the dropdown menu?

No further information beyond this:

![QQ截图20241118234340](https://github.com/user-attachments/assets/c0d2b4d3-c505-4a18-8f57-7f72a426433a)

And my system version is `Windows11 22H2`.

**Comment by fhammerschmidt** [comment]
I can reproduce the error in your output if I do 
`pnpm run res:clean`
and do not trigger a build afterwards.

But as soon as I do 
`pnpm run res:dev`

Everything works again. And to me it really just looks like the LSP got stuck and needs a restart. Btw. when you close all rescript tabs and open one again, it also gets restarted.

**Comment by BIYUEHU** [comment]
> I can reproduce the error in your output if I do `pnpm run res:clean` and do not trigger a build afterwards.
> 
> But as soon as I do `pnpm run res:dev`
> 
> Everything works again. And to me it really just looks like the LSP got stuck and needs a restart. Btw. when you close all rescript tabs and open one again, it also gets restarted.

I've tried to clean rescript and restart VSC, but it has no effect.

After I  closed `Incremental Typechecking: Enabled` at settings, the problem seems to be solved:

![QQ截图20241120121317](https://github.com/user-attachments/assets/5ca545bf-02f4-4be0-9e0d-82b2f98a431b)

Now I need to run `res:dev` all the time when developing rescript before there is type checking (`LSP` of the extension doesn't seem to be working), is this normal?

**Comment by zth** [comment]
There might very well be some issue with windows and path handling for incremental typechecking, yes.

**Comment by fhammerschmidt** [comment]
Not so fast, GitHub.

@BIYUEHU can you try setting your extension to the pre-release version 1.59.6 and see if incremental compilation works there?

**Comment by BIYUEHU** [comment]
> Not so fast, GitHub.
> 
> @BIYUEHU can you try setting your extension to the pre-release version 1.59.6 and see if incremental compilation works there?

I tried.Type checking is working normally:

![QQ截图20241122121319](https://github.com/user-attachments/assets/4f5f3f09-19b4-4708-9007-343494607835)

But the first problem still exists:

![QQ截图20241122121347](https://github.com/user-attachments/assets/2745c5ec-748e-4f8a-bb77-611e5aef15e0)

**Comment by Naganamin** [comment]
Confirm to have same issue on windows. It works fine on WSL 2

**Comment by snatvb** [comment]
this is LSP Issue, on neovim the same issue :(

**Comment by sainthkh** [comment]
I'm still having this problem on Windows with `1.62.0`. It's the same problem.

**Comment by nojaf** [comment]
I believe I have a fix for this, could someone try out the `vsix` on https://github.com/rescript-lang/rescript-vscode/actions/runs/17259865998?pr=1120

**Comment by snatvb** [comment]
@nojaf  same issue in neovim, that's not vscode ext prbolem, bug in lsp

**Comment by nojaf** [comment]
@snatvb my fix is in LSP server, should work for neovim as well.

**Comment by ssmfd** [comment]
@nojaf Okay, so, here they are:

<details>
<summary><code>build.ninja</code> contents</summary>

```
rescript = 1
g_finger := /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/@rescript/core/lib/ocaml/install.stamp
rule astj
  command = /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/rescript/linux/bsc.exe  -bs-v 11.1.4 -uncurried -open RescriptCore -absname -bs-ast -o $out $i
o src/Demo.ast : astj ../../src/Demo.res
rule deps
  command = /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/rescript/linux/bsb_helper.exe -hash 13bda1cb041571eac974df229b45b3be $in
  restat = 1
o src/Demo.d : deps src/Demo.ast
rule mij
  command = /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/rescript/linux/bsc.exe -I src/lib/oak -I src/lib/web-api -I src/lib -I src/routers -I src/utils -I src -I /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/@rescript/core/lib/ocaml -open RescriptCore  -uncurried -bs-package-name server-rescript -bs-package-output esmodule:$in_d:.res.mjs -bs-v $g_finger $i
  dyndep = 1
  restat = 1
o src/Demo.cmj src/Demo.cmi ../../src/Demo.res.mjs : mij src/Demo.ast
o src/Main.ast : astj ../../src/Main.res
o src/Main.d : deps src/Main.ast
o src/Main.cmj src/Main.cmi ../../src/Main.res.mjs : mij src/Main.ast
o src/Types.ast : astj ../../src/Types.res
o src/Types.d : deps src/Types.ast
o src/Types.cmj src/Types.cmi ../../src/Types.res.mjs : mij src/Types.ast
o src/AppState.ast : astj ../../src/AppState.res
o src/AppState.d : deps src/AppState.ast
o src/AppState.cmj src/AppState.cmi ../../src/AppState.res.mjs : mij src/AppState.ast
o src/utils/HttpError.ast : astj ../../src/utils/HttpError.res
o src/utils/HttpError.d : deps src/utils/HttpError.ast
o src/utils/HttpError.cmj src/utils/HttpError.cmi ../../src/utils/HttpError.res.mjs : mij src/utils/HttpError.ast
o src/utils/DataValidation.ast : astj ../../src/utils/DataValidation.res
o src/utils/DataValidation.d : deps src/utils/DataValidation.ast
o src/utils/DataValidation.cmj src/utils/DataValidation.cmi ../../src/utils/DataValidation.res.mjs : mij src/utils/DataValidation.ast
o src/routers/SessionsRouter.ast : astj ../../src/routers/SessionsRouter.res
o src/routers/SessionsRouter.d : deps src/routers/SessionsRouter.ast
o src/routers/SessionsRouter.cmj src/routers/SessionsRouter.cmi ../../src/routers/SessionsRouter.res.mjs : mij src/routers/SessionsRouter.ast
o src/lib/web-api/Headers.ast : astj ../../src/lib/web-api/Headers.res
o src/lib/web-api/Headers.d : deps src/lib/web-api/Headers.ast
o src/lib/web-api/Headers.cmj src/lib/web-api/Headers.cmi ../../src/lib/web-api/Headers.res.mjs : mij src/lib/web-api/Headers.ast
o src/lib/web-api/FormData.ast : astj ../../src/lib/web-api/FormData.res
o src/lib/web-api/FormData.d : deps src/lib/web-api/FormData.ast
o src/lib/web-api/FormData.cmj src/lib/web-api/FormData.cmi ../../src/lib/web-api/FormData.res.mjs : mij src/lib/web-api/FormData.ast
o src/lib/oak/Router.ast : astj ../../src/lib/oak/Router.res
o src/lib/oak/Router.d : deps src/lib/oak/Router.ast
o src/lib/oak/Router.cmj src/lib/oak/Router.cmi ../../src/lib/oak/Router.res.mjs : mij src/lib/oak/Router.ast
o src/lib/oak/Context.ast : astj ../../src/lib/oak/Context.res
o src/lib/oak/Context.d : deps src/lib/oak/Context.ast
o src/lib/oak/Context.cmj src/lib/oak/Context.cmi ../../src/lib/oak/Context.res.mjs : mij src/lib/oak/Context.ast
o src/lib/oak/Cookies.ast : astj ../../src/lib/oak/Cookies.res
o src/lib/oak/Cookies.d : deps src/lib/oak/Cookies.ast
o src/lib/oak/Cookies.cmj src/lib/oak/Cookies.cmi ../../src/lib/oak/Cookies.res.mjs : mij src/lib/oak/Cookies.ast
o src/lib/oak/Request.ast : astj ../../src/lib/oak/Request.res
o src/lib/oak/Request.d : deps src/lib/oak/Request.ast
o src/lib/oak/Request.cmj src/lib/oak/Request.cmi ../../src/lib/oak/Request.res.mjs : mij src/lib/oak/Request.ast
o src/lib/oak/Response.ast : astj ../../src/lib/oak/Response.res
o src/lib/oak/Response.d : deps src/lib/oak/Response.ast
o src/lib/oak/Response.cmj src/lib/oak/Response.cmi ../../src/lib/oak/Response.res.mjs : mij src/lib/oak/Response.ast
o src/lib/oak/HttpStatus.ast : astj ../../src/lib/oak/HttpStatus.res
o src/lib/oak/HttpStatus.d : deps src/lib/oak/HttpStatus.ast
o src/lib/oak/HttpStatus.cmj src/lib/oak/HttpStatus.cmi ../../src/lib/oak/HttpStatus.res.mjs : mij src/lib/oak/HttpStatus.ast
o src/lib/oak/Middleware.ast : astj ../../src/lib/oak/Middleware.res
o src/lib/oak/Middleware.d : deps src/lib/oak/Middleware.ast
o src/lib/oak/Middleware.cmj src/lib/oak/Middleware.cmi ../../src/lib/oak/Middleware.res.mjs : mij src/lib/oak/Middleware.ast
o src/lib/oak/Application.ast : astj ../../src/lib/oak/Application.res
o src/lib/oak/Application.d : deps src/lib/oak/Application.ast
o src/lib/oak/Application.cmj src/lib/oak/Application.cmi ../../src/lib/oak/Application.res.mjs : mij src/lib/oak/Application.ast
o src/lib/oak/RequestBody.ast : astj ../../src/lib/oak/RequestBody.res
o src/lib/oak/RequestBody.d : deps src/lib/oak/RequestBody.ast
o src/lib/oak/RequestBody.cmj src/lib/oak/RequestBody.cmi ../../src/lib/oak/RequestBody.res.mjs : mij src/lib/oak/RequestBody.ast
```

</details>

<details>
<summary><code>ReScript Language Server</code> logs</summary>

```
About to invoke bsc from "d:\prog\fandb\server-rescript", used bsb
d:\prog\fandb\server-rescript\node_modules\rescript\win32\bsc.exe "-I" "d:\prog\fandb\server-rescript\lib\bs\src\lib\oak" "-I" "d:\prog\fandb\server-rescript\lib\bs\src\lib\web-api" "-I" "d:\prog\fandb\server-rescript\lib\bs\src\lib" "-I" "d:\prog\fandb\server-rescript\lib\bs\src\routers" "-I" "d:\prog\fandb\server-rescript\lib\bs\src\utils" "-I" "d:\prog\fandb\server-rescript\lib\bs\src" "-I" "d:\mnt\C878A2AF78A29C26\prog\fandb\server-rescript\node_modules\@rescript\core\lib\ocaml" "-open" "RescriptCore" "-uncurried" "-bs-package-name" "server-rescript" "-bs-v" "1756306508635" "-color" "never" "-ignore-parse-errors" "d:\prog\fandb\server-rescript\lib\bs\___incremental\Middleware.res"
Recompiled Middleware.res in 0.026784299999999348s
Resetting compilation status.
```

</details>

**Comment by nojaf** [comment]
@ssmfd are you doing WSL things?

I see both

```
rule mij
  command = /mnt/C878A2AF78A29C26/prog/fandb/server-rescript/node_modules/rescript/linux/bsc.exe
```

and

`"d:\mnt\C878A2AF78A29C26\prog\fandb\server-rescript\node_modules\@rescript\core\lib\ocaml"`

So, I wonder where this mix is coming from?

My fix of today, does have a "is native windows" check, so if you are not on puresang Windows, it might not be helping you situation.

**Comment by nojaf** [comment]
@ssmfd it is worth posting the same info I asked last time.
Also remove the `lib` folder and do a build before starting the IDE.
Just to confirm nothing was lingering.

**Comment by nojaf** [comment]
> I still get errors.

Could you be more specific, what exactly do you get?
Can you provide a sample repo, I did some more testing on Windows and all did seem to work.
One thing I noticed is that, certain things are not picked up like installing a new package without a `rescript build` will not work. But that is not Windows specific I believe.

I merged my PR as I do believe it is a step in the right direction.
Other people can try out the prerelease version.

**Comment by ssmfd** [comment]
Okay, here's a minimum sample repo: https://github.com/ssmfd/rescript-issue-1056-sample

After testing for a bit, I think the problem is with modules that are in subfolders. In the example I use a type from a module in the same folder, and another type from a module that's in a subfolder, and that's the one that fails.

The compiler finds all of the modules and compiles them with no errors, but VSCode still marks variable `b`'s type as not found.

**Comment by nojaf** [comment]
Thanks, will take a look tomorrow!

**Comment by nojaf** [comment]
Great, thanks for confirming!
I've merged that last fix, so shortly it will become available in the next prerelease.
If someone else could give it a go, we can indeed close this issue and prepare a new stable release of the extension.

**Comment by nojaf** [comment]
I'm going to close this issue, if anyone has related problems, please do not hesitate to open a new issue.
